### PR TITLE
move-only: move coins statistics utils out of RPC

### PR DESCRIPTION
**Pull request by jamesob**
This is part of the [assumeutxo project](https://github.com/bitcoin/bitcoin/projects/11):

Parent PR: #15606 
Issue: #15605 
Specification: https://github.com/jamesob/assumeutxo-docs/tree/master/proposal

---

In the short-term, this move-only commit will help with fuzzing (https://github.com/bitcoin/bitcoin/pull/15606#issuecomment-524482297). Later, these procedures will be used to compute statistics (particularly a content hash) for UTXO sets coming in from snapshots.

Most easily reviewed with `git ... --color-moved=dimmed_zebra`. A nice follow-up would be adding unittests, which I'll do if nobody else gets around to it.

### CODE REVIEW

**Review by maflcko — COMMENTED** [review]
Concept ACK

**Review by ryanofsky — APPROVED** [review]
utACK 07da7c422d875e54176a81b37de011a4761966cc

**Review by ariard — APPROVED** [review]
Reviewed and tested ACK 8f394b5

File: src/node/coinstats.cpp [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+{
+    std::unique_ptr<CCoinsViewCursor> pcursor(view->Cursor());
+    assert(pcursor);
+
+    CHashWriter ss(SER_GETHASH, PROTOCOL_VERSION);
+    stats.hashBlock = pcursor->GetBestBlock();
+    {
+        LOCK(cs_main);
+        CBlockIndex* block = LookupBlockIndex(stats.hashBlock);
+        stats.nHeight = block ? block->nHeight : -1;
```
Comment (Sjors) [inline src/node/coinstats.cpp]: This change deserves its own commit?

**Review by Sjors — APPROVED** [review]
ACK 8f394b5809a5ccf042e7ddf4d3921acb3a7a67a2. Happy to see more RPC code move to a seperate and reusable spot. `gettxoutsetinfo` still works.

File: src/node/coinstats.cpp (hunk 1 above)
Comment (promag) [reply src/node/coinstats.cpp]: Agree with the separate commit.

File: src/node/coinstats.h [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    uint64_t nBogoSize;
+    uint256 hashSerialized;
+    uint64_t nDiskSize;
+    CAmount nTotalAmount;
+
+    CCoinsStats() : nHeight(0), nTransactions(0), nTransactionOutputs(0), nBogoSize(0), nDiskSize(0), nTotalAmount(0) {}
+};
+
+//! Calculate statistics about the unspent transaction output set
+bool GetUTXOStats(CCoinsView *view, CCoinsStats &stats);
```
Comment (promag) [inline src/node/coinstats.h]: Could fix format in this new line (space before arg name).

File: src/node/coinstats.h [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+#include <amount.h>
+#include <sync.h>
+#include <uint256.h>
+
+
+class CCoinsView;
+
+struct CCoinsStats
+{
+    int nHeight;
```
[the remark this answers is not in the document]
Comment (promag) [reply src/node/coinstats.h]: If you are going to push a separate commit then you could also drop the constructor and initialize members in place, like `int nHeight{0};`.

File: src/node/coinstats.h [hunk 4]
```diff
@@ -0,0 +1,33 @@
+// Copyright (c) 2010 Satoshi Nakamoto
+// Copyright (c) 2009-2019 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_NODE_COINSTATS_H
+#define BITCOIN_NODE_COINSTATS_H
+
+#include <amount.h>
+#include <sync.h>
```
Comment (promag) [inline src/node/coinstats.h]: What is this for?

File: src/node/coinstats.h [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+#include <uint256.h>
+
+
+class CCoinsView;
+
+struct CCoinsStats
+{
+    int nHeight;
+    uint256 hashBlock;
+    uint64_t nTransactions;
```
Comment (promag) [inline src/node/coinstats.h]: Should `#include <cstdint>`.

**Review by promag — COMMENTED** [review]
Concept ACK, verified it's almost moved-only.

File: src/node/coinstats.cpp [hunk 6]
```diff
@@ -0,0 +1,77 @@
+// Copyright (c) 2010 Satoshi Nakamoto
+// Copyright (c) 2009-2019 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <node/coinstats.h>
+
+#include <amount.h>
+#include <coins.h>
+#include <chain.h>
```
Comment (promag) [inline src/node/coinstats.cpp]: nit sort :trollface:

**Review by maflcko — APPROVED** [review]

**Comment by promag** [comment]
ACK
